File: database-driver-plugin/src/main/java/de/lino/database/database/nosql/redis/RedisDatabaseProvider.java
package de.lino.database.database.nosql.redis;

import de.lino.database.database.AbstractCachedDatabaseSection;
import de.lino.database.database.AbstractLazyDatabaseProvider;
import de.lino.database.database.SectionConfig;
import de.lino.database.database.auth.Credentials;
import de.lino.database.database.DatabaseProvider;
import de.lino.database.database.DatabaseSection;
import de.lino.database.database.notification.RedisCounterService;
import org.jetbrains.annotations.NotNull;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;
import redis.clients.jedis.params.ScanParams;
import redis.clients.jedis.resps.ScanResult;

import java.util.List;
import java.util.function.Consumer;

/**
 * The {@link DatabaseProvider} backed by a Redis database, each {@link DatabaseSection} a
 * {@code "<name>:*"} key prefix via {@link RedisDatabaseSection}, all sharing this database's
 * single {@link JedisPool}. Section lifecycle and caching live in
 * {@link AbstractLazyDatabaseProvider}; this class only supplies the keyspace-level storage
 * operations - discovering prefixes, constructing a {@link RedisDatabaseSection}, wiping a
 * prefix. {@link JedisPool} is itself thread-safe and designed for concurrent multi-threaded
 * use, so every method here is safe to call concurrently without additional locking.
 */
public class RedisDatabaseProvider extends AbstractLazyDatabaseProvider {

    /**
     * How many keys each {@code SCAN} round trip asks the server for - the historical batch
     * size, bounding a scan's per-round-trip work without starving it.
     */
    private static final int SCAN_BATCH_SIZE = 100;

    /**
     * The connection pool shared by this database and every {@link RedisDatabaseSection} it creates.
     */
    private final JedisPool jedisPool;

    /**
     * Lazily constructed by {@link #counterService()}; {@code volatile} plus double-checked
     * locking there so concurrent first-callers can't each construct and race to publish their
     * own instance - both cheap to guard against and worth guarding, since every caller must end
     * up sharing the exact same {@link JedisRedisCounterService}, not one each.
     */
    private volatile RedisCounterService counterService;

    /**
     * Connects to a Redis database with {@code credentials} and discovers every existing key
     * prefix as a section name. Only names, from a single keyspace {@code SCAN} - no section
     * objects, no values - so construction cost is O(keys) once. Historically this constructor
     * created one section <em>per key</em> (not per prefix), each of which then ran its own
     * full-keyspace scan - O(keys²) work that degraded badly with key count, and section names
     * that were really key names; both are gone with prefix discovery.
     *
     * @param credentials the login credentials and connection details to connect with
     */
    public RedisDatabaseProvider(@NotNull Credentials credentials) {

        final JedisPoolConfig jedisPoolConfig = new JedisPoolConfig();
        jedisPoolConfig.setMaxTotal(50);
        jedisPoolConfig.setMaxIdle(10);
        jedisPoolConfig.setMinIdle(2);
        jedisPoolConfig.setTestOnBorrow(true);

        if (credentials.getUserName().isEmpty() && credentials.getPassword().isEmpty()) {
            this.jedisPool = new JedisPool(jedisPoolConfig, credentials.getAddress(), credentials.getPort());
            try (final Jedis jedis = this.jedisPool.getResource()) {
                jedis.select(Integer.parseInt(credentials.getDatabase()));
            }
        } else {
            this.jedisPool = new JedisPool(jedisPoolConfig, "redis://:" + credentials.getPassword() + "@" + credentials.getAddress() + ":" + credentials.getPort() + "/" + credentials.getDatabase());
        }

        this.reload();

    }

    @Override
    public void shutdown() {
        this.jedisPool.close();
        this.forgetSections();
    }

    /**
     * {@inheritDoc}
     * <p>
     * A single cursor-based {@code SCAN} over the whole keyspace, reducing each key
     * {@code "<prefix>:<id>"} to its prefix (a key without a {@code ':'} passes through
     * whole, mirroring how {@link RedisDatabaseSection} would name it). Duplicates collapse in
     * the caller's name set, so N keys cost one O(N) pass - not the historical
     * one-scan-per-key O(N²).
     */
    @Override
    protected void discoverNames(@NotNull Consumer<String> consumer) {

        String cursor = "0";
        final ScanParams scanParams = new ScanParams().match("*").count(SCAN_BATCH_SIZE);

        try (final Jedis jedis = this.jedisPool.getResource()) {

            do {

                final ScanResult<String> result = jedis.scan(cursor, scanParams);

                for (final String key : result.getResult()) {
                    final int separator = key.indexOf(':');
                    consumer.accept(separator < 0 ? key : key.substring(0, separator));
                }

                cursor = result.getCursor();

            } while (!cursor.equals("0"));

        }

    }

    @Override
    protected AbstractCachedDatabaseSection constructSection(@NotNull String name, @NotNull SectionConfig config) {
        return new RedisDatabaseSection(this.jedisPool, name, config);
    }

    /**
     * {@inheritDoc}
     * <p>
     * A cursor-based {@code SCAN} over {@code "<name>:*"} with one {@code DEL} per scanned
     * batch rather than one per key, cutting round trips from O(matched keys) to
     * O(matched keys / {@value #SCAN_BATCH_SIZE}). The pattern deliberately includes the
     * {@code ':'} separator so deleting section {@code "users"} can never take keys of an
     * unrelated section that merely shares the character prefix (like {@code "users2"}) with
     * it.
     */
    @Override
    protected void dropSectionRemote(@NotNull String name) {

        try (final Jedis jedis = this.jedisPool.getResource()) {

            String cursor = "0";
            final ScanParams scanParams = new ScanParams().match(name + ":*").count(SCAN_BATCH_SIZE);

            do {

                final ScanResult<String> result = jedis.scan(cursor, scanParams);
                final List<String> keys = result.getResult();

                if (!keys.isEmpty()) jedis.del(keys.toArray(new String[0]));

                cursor = result.getCursor();

            } while (!cursor.equals("0"));

        }

    }

    /**
     * Returns this provider's {@link RedisCounterService}, constructing it on first call and
     * reusing that same instance afterward. The returned service shares {@link #jedisPool} with
     * every {@link RedisDatabaseSection} this provider manages, rather than opening a second,
     * independent {@link JedisPool} against the same Redis instance.
     *
     * @return this provider's lazily-constructed, shared {@link RedisCounterService}
     */
    public @NotNull RedisCounterService counterService() {

        RedisCounterService service = this.counterService;
        if (service != null) return service;

        synchronized (this) {
            service = this.counterService;
            if (service == null) {
                service = new JedisRedisCounterService(this.jedisPool);
                this.counterService = service;
            }
        }

        return service;

    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/nosql/redis/redis_database_provider.py
"""Mirror of ``de.lino.database.database.nosql.redis.RedisDatabaseProvider``."""

from __future__ import annotations

import threading
from collections.abc import Callable
from typing import Any

from database_driver.api.database.auth.credentials import Credentials
from database_driver.api.database.notification.redis_counter_service import RedisCounterService
from database_driver.api.database.section_config import SectionConfig

from database_driver.plugin.database.abstract_cached_database_section import AbstractCachedDatabaseSection
from database_driver.plugin.database.abstract_lazy_database_provider import AbstractLazyDatabaseProvider
from database_driver.plugin.database.nosql.redis.redis_database_section import RedisDatabaseSection

_SCAN_BATCH_SIZE = 100
"""How many keys each ``SCAN`` round trip asks the server for - the historical batch
size, bounding a scan's per-round-trip work without starving it."""


class RedisDatabaseProvider(AbstractLazyDatabaseProvider):
    """The ``DatabaseProvider`` backed by a Redis database, each section a ``"<name>:*"``
    key prefix via ``RedisDatabaseSection``, all sharing this database's single pooled
    redis-py client. Section lifecycle and caching live in
    ``AbstractLazyDatabaseProvider``; this class only supplies the keyspace-level
    storage operations - discovering prefixes, constructing a section, wiping a prefix.
    redis-py's ``Redis`` client is itself pool-backed and thread-safe, so every method
    here is safe to call concurrently without additional locking.

    Discovers every existing key prefix as a section name from a single keyspace
    ``SCAN`` - O(keys) once. (The historical Java implementation once created one
    section *per key*, each running its own full-keyspace scan - O(keys²) work; both are
    long gone with prefix discovery, here as there.)
    """

    def __init__(self, credentials: Credentials) -> None:
        super().__init__()

        import redis

        if not credentials.user_name and not credentials.password:
            self.client: Any = redis.Redis(
                host=credentials.address,
                port=credentials.port,
                db=int(credentials.database),
                max_connections=50,
            )
        else:
            self.client = redis.Redis.from_url(
                f"redis://:{credentials.password}@{credentials.address}:{credentials.port}/{credentials.database}",
                max_connections=50,
            )

        # Lazily constructed by counter_service(); double-checked so concurrent
        # first-callers all end up sharing the exact same instance, not one each.
        self._counter_service: RedisCounterService | None = None
        self._counter_lock = threading.Lock()

        self.reload()

    def shutdown(self) -> None:
        self.client.close()
        self.forget_sections()

    def discover_names(self, consumer: Callable[[str], None]) -> None:
        """A single cursor-based ``SCAN`` over the whole keyspace, reducing each key
        ``"<prefix>:<id>"`` to its prefix (a key without a ``':'`` passes through whole,
        mirroring how ``RedisDatabaseSection`` would name it). Duplicates collapse in the
        caller's name set, so N keys cost one O(N) pass."""
        for key in self.client.scan_iter(match="*", count=_SCAN_BATCH_SIZE):
            key_text = key.decode("utf-8") if isinstance(key, bytes) else key
            separator = key_text.find(":")
            consumer(key_text if separator < 0 else key_text[:separator])

    def construct_section(self, name: str, config: SectionConfig) -> AbstractCachedDatabaseSection:
        return RedisDatabaseSection(self.client, name, config)

    def drop_section_remote(self, name: str) -> None:
        """A cursor-based ``SCAN`` over ``"<name>:*"`` with one ``DEL`` per scanned batch
        rather than one per key. The pattern deliberately includes the ``':'`` separator
        so deleting section ``"users"`` can never take keys of an unrelated section that
        merely shares the character prefix (like ``"users2"``) with it."""
        batch: list[Any] = []
        for key in self.client.scan_iter(match=f"{name}:*", count=_SCAN_BATCH_SIZE):
            batch.append(key)
            if len(batch) >= _SCAN_BATCH_SIZE:
                self.client.delete(*batch)
                batch.clear()
        if batch:
            self.client.delete(*batch)

    def counter_service(self) -> RedisCounterService:
        """Returns this provider's ``RedisCounterService``, constructing it on first call
        and reusing that same instance afterward. The returned service shares this
        provider's own client (and so its connection pool) with every section this
        provider manages, rather than opening a second, independent pool against the
        same Redis instance."""
        service = self._counter_service
        if service is not None:
            return service

        with self._counter_lock:
            service = self._counter_service
            if service is None:
                from database_driver.plugin.database.nosql.redis.redis_py_counter_service import (
                    RedisPyCounterService,
                )

                service = RedisPyCounterService(self.client)
                self._counter_service = service

        return service
